sw/device/lib/dif/dif_i2c.h

```diff
@@ -0,0 +1,390 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_DIF_DIF_I2C_H_
+#define OPENTITAN_SW_DEVICE_LIB_DIF_DIF_I2C_H_
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Represents a speed setting for an I2C component: standard, fast, and
+ * fast plus, corresponding to 100 kbaud, 400 kbaud, and 1 Mbaud,
+ * respectively.
+ */
+typedef enum dif_i2c_speed {
+  /**
+   * Standard speed, 100 kilobaud.
+   */
+  kDifI2cSpeedStandard,
+  /**
+   * Fast speed, 400 kilobaud.
+   */
+  kDifI2cSpeedFast,
+  /**
+   * Fast plus speed, 1 megabaud.
+   */
+  kDifI2cSpeedFastPlus,
+} dif_i2c_speed_t;
+
+/*
+ * Timing parameters for an I2C device. Actually computing these values is
+ * somewhat complicated, so this struct should be assembled using the
+ * `dif_i2c_compute_timing()` function. A caller is, however, free to assemble
+ * this struct or modify a return value of `dif_i2c_compute_timing()`, so long
+ * as the spec is respected.
+ *
+ * These values correspond to those in Table 10 of the I2C spec, and are given
+ * in units of input clock cycles.
+ */
+typedef struct dif_i2c_timing_params {
```

Comment:
We like this interface because having the DIF provide some defaults is good, but being able to tweak the parameters in drivers later will be useful for getting finer-grained control.